packages/core/contracts/financial-templates/optimistic-rewarder/OptimisticRewarder.sol

```diff
@@ -0,0 +1,429 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+pragma solidity ^0.8.0;
+
+import "@openzeppelin/contracts/token/ERC721/ERC721.sol";
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+
+import "../../common/implementation/Lockable.sol";
+import "../../common/implementation/Testable.sol";
+import "../../oracle/interfaces/FinderInterface.sol";
+import "../../oracle/interfaces/StoreInterface.sol";
+import "../../oracle/interfaces/IdentifierWhitelistInterface.sol";
+import "../../oracle/implementation/Constants.sol";
+import "../../oracle/interfaces/SkinnyOptimisticOracleInterface.sol";
+import "../../common/implementation/AncillaryData.sol";
+import "../../common/interfaces/AddressWhitelistInterface.sol";
+import "./OptimisticRewarderToken.sol";
+
+/**
+ * @notice The base rewarder contract. This manages depositing rewards and paying them out to tokenholders using values
+ * backed by the OptimisticOracle's dispute process.
+ */
+abstract contract OptimisticRewarderBase is Lockable {
+    using SafeERC20 for IERC20;
+
+    struct RedemptionAmount {
+        uint256 amount;
+        IERC20 token;
+    }
+
+    struct Redemption {
+        uint256 finalFee;
+        uint256 expiryTime;
+    }
+
+    // Constants.
+    FinderInterface public finder;
+    uint256 public liveness;
+    uint256 public bond;
+    IERC20 public bondToken;
+    bytes32 public identifier;
+    bytes public customAncillaryData;
+
+    // Mapping to track redemptions.
+    mapping(bytes32 => Redemption) public redemptions;
+
+    // Mapping to track the past total cumulative redemptions for tokenIds.
+    mapping(uint256 => mapping(IERC20 => uint256)) public redeemedAmounts;
+
+    // This allows other contracts to publish reward updates.
+    event UpdateToken(uint256 indexed tokenId, address indexed caller, bytes data);
```

Comment:
So are Redemption attempts (i.e. Submissions) linked to `UpdateToken` events? And are disputes essentially checking whether the history of `UpdateToken` events produces the proposed `cumulativeRedemptions` amounts?

Reply:
This is where the contract is really flexible, but admittedly a little confusing. 

UpdateToken events are an _easy way_ to track interactions that affect the token's rewards. However, the identifier could be specified s.t. it relies on data that is logged completely outside of this contract.

I wanted a simple way to log _something_, but it seems overly restrictive to _only allow_ UpdateToken events to affect the redemption value.

Thoughts?

Reply:
> I wanted a simple way to log something, but it seems overly restrictive to only allow UpdateToken events to affect the redemption value.

I agree its overly restrictive to only rely on UpdateToken events. I guess I'm not fully understanding how this would work in practice

Reply:
The update token or the events that occur outside the contract? I can give an example of either.